ehrapy/preprocessing/_data_imputation.py

```diff
@@ -386,6 +386,283 @@ def miss_forest_impute(
+
+    See https://github.com/iskandr/fancyimpute/blob/master/fancyimpute/soft_impute.py
+    Matrix completion by iterative soft thresholding of SVD decompositions.
+
+    Args:
+        adata: The AnnData object to use SoftImpute on.
+        var_names: A list of var names indicating which columns to impute (if None -> all columns).
+        copy: Whether to return a copy or act in place.
+        warning_threshold: Threshold of percentage of missing values to display a warning for (default: 30).
+        shrinkage_value : Value by which we shrink singular values on each iteration. If omitted then the default value will be the maximum singular value of the initialized matrix (zeros for missing values) divided by 50.
+        convergence_threshold : Minimum ration difference between iterations (as a fraction of the Frobenius norm of the current solution) before stopping.
+        max_iters: Maximum number of SVD iterations.
+        max_rank: Perform a truncated SVD on each iteration with this value as its rank.
+        n_power_iterations: Number of power iterations to perform with randomized SVD.
+        init_fill_method: How to initialize missing values of data matrix, default is to fill them with zeros.
+        min_value: Smallest allowable value in the solution.
+        max_value: Largest allowable value in the solution.
+        normalizer: Any object (such as BiScaler) with fit() and transform() methods.
+        verbose: Print debugging info.
+
+    Returns:
+        The imputed (but unencoded) AnnData object
+
+    Example:
+        .. code-block:: python
+
+            import ehrapy as ep
+
+            adata = ep.dt.mimic_2(encoded=True)
+            ep.pp.soft_impute(adata)
+    """
+    if copy:
+        adata = adata.copy()
+
+    _warn_imputation_threshold(adata, var_names, threshold=warning_threshold)
+
+    if check_module_importable("fancyimpute"):
+        from fancyimpute import SoftImpute
+
+        SoftImpute()
+    else:
+        print("[bold yellow]fancyimpute is not available. Install via [blue]pip install fancyimpute.")
+
+    with Progress(
+        "[progress.description]{task.description}",
+        SpinnerColumn(),
+        refresh_per_second=1500,
+    ) as progress:
+        progress.add_task("[blue]Running SoftImpute", total=1)
+        if np.issubdtype(adata.X.dtype, np.number):
+            _soft_impute(
+                adata,
+                var_names,
+                shrinkage_value,
+                convergence_threshold,
+                max_iters,
+                max_rank,
+                n_power_iterations,
+                init_fill_method,
+                min_value,
+                max_value,
+                normalizer,
+                verbose,
+            )
+        else:
+            # ordinal encoding is used since non-numerical data can not be imputed using SoftImpute
+            enc = OrdinalEncoder()
+            adata.X = enc.fit_transform(adata.X)
```

Comment:
IMO, we should restrict encoding here to all non numerical columns (stored in `uns["non_numerical_columns"]`). Guess this is something that should be done in `knn_impute` as well.

Reply:
I understand, only encode non numerical columns. Can you please have a look if this ok? I tested this worked well.
```
column_indices = get_column_indices(adata, adata.uns['non_numerical_columns'])
enc = OrdinalEncoder()
adata.X[::, column_indices] = enc.fit_transform(adata.X[::, column_indices])
impute(...)
adata.X[::, column_indices] = enc.inverse_transform(adata.X[::, column_indices])
```

Reply:
LGTM